pkg/kubectl/rolling_updater.go

```diff
@@ -93,27 +79,436 @@ const (
+// duration of the update. Each interval, the updater will scale down whatever
+// it can without violating the minimum, and will scale up as much as it can
+// up to a maximum increment. This means amount scaled up or down each
+// interval will vary based on the timeliness of readiness and the updater
+// will always try to make progress, even slowly.
+//
+// If an update from newRc to oldRc is already in progress, we attempt to
+// drive it to completion. If an error occurs at any step of the update, the
+// error will be returned.
 //
-// After the successful scale-up of the controller, the controller is given to
-// the UpdateAcceptor. If the UpdateAcceptor rejects the controller, the
-// deployment is stopped with an error.
-type UpdateAcceptor interface {
-	// Accept returns nil if the controller is okay, otherwise returns an error.
-	Accept(*api.ReplicationController) error
+// TODO: make this handle performing a rollback of a partially completed
+// rollout.
+func (r *RollingUpdater) Update(config *RollingUpdaterConfig) error {
+	out := config.Out
+	oldRc := config.OldRc
+
+	// Find an existing controller (for continuing an interrupted update) or
+	// create a new one if necessary.
+	sourceId := fmt.Sprintf("%s:%s", oldRc.Name, oldRc.UID)
+	newRc, existed, err := r.getOrCreateTargetController(config.NewRc, sourceId)
+	if err != nil {
+		return err
+	}
+	if existed {
+		fmt.Fprintf(out, "Continuing update with existing controller %s.\n", newRc.Name)
+	} else {
+		fmt.Fprintf(out, "Created %s\n", newRc.Name)
+	}
+	// Extract the desired replica count from the controller.
+	desired, err := strconv.Atoi(newRc.Annotations[desiredReplicasAnnotation])
+	if err != nil {
+		return fmt.Errorf("Unable to parse annotation for %s: %s=%s",
+			newRc.Name, desiredReplicasAnnotation, newRc.Annotations[desiredReplicasAnnotation])
+	}
+
+	// This is the maximum number of replicas we'll scale in either direction.
+	maxIncrement := 1
+	// This is the minimum number of replicas (total of old/new) which must
+	// remain ready for the duration of the update if progress is to be made.
+	minReady := oldRc.Spec.Replicas - maxIncrement
+	// This drives the scaling direction. The default is scale up/down. If
+	// false, use down/up (in-place).
+	scaleUpFirst := true
+	// By default, use an availability check which is based on pod existence to
+	// avoid waiting for readiness to make progress.
+	availabilityCheck := createdCheck
+	if config.UpdatePercent != nil {
+		percent := *config.UpdatePercent
+		// Compute the scale increment based on a percentage of the new desired
+		// count.
+		maxIncrement = int(math.Ceil(float64(desired) * (math.Abs(float64(percent)) / 100)))
+		// Compute the minimum ready requirement as a percentage of the old
+		// replica count.
+		percentOld := int(math.Ceil(float64(oldRc.Spec.Replicas) * (math.Abs(float64(percent)) / 100)))
+		minReady = oldRc.Spec.Replicas - percentOld
```

Comment:
Subtly different: The max increment is based on a percentage of the *new desired count*, while the minimum requirement is based on a percentage of the *old replica count*. Sometimes these will be the same, if new desired is the same as the old spec count (if the old/new replica counts match), but we must also support asymmetrical updates. For example, old replicas is 2 and  new desired is 6 with a 50% update, using new as the minimum basis would mean our minimum is 3 (which means our old/current deployment is already nonsensically violating the minimum).

Does that make sense? If so I should probably make a better comment in the code.